test/extended/util/disruption/controlplane/termination.go

```diff
@@ -0,0 +1,120 @@
+package controlplane
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"k8s.io/apimachinery/pkg/runtime"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/tools/cache"
+
+	"k8s.io/client-go/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+
+	kwatch "k8s.io/apimachinery/pkg/watch"
+
+	"k8s.io/client-go/tools/watch"
+
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/upgrades"
+)
+
+// GracefulTerminationTest watches events of non-graceful terminations of kube-apiserver.
+type GracefulTerminationTest struct{}
+
+func (GracefulTerminationTest) Name() string { return "kube-apiserver-terminates-gracefully" }
+func (GracefulTerminationTest) DisplayName() string {
+	return "Kubernetes API server terminates gracefully"
+}
+func (GracefulTerminationTest) Setup(f *framework.Framework)    {}
+func (GracefulTerminationTest) Teardown(f *framework.Framework) {}
+func (t *GracefulTerminationTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer cancel()
+		<-done
+	}()
+
+	m := EventMonitor{
+		Namespace: "openshift-kube-apiserver",
+		Filter: func(ev *corev1.Event) bool {
+			return ev.Reason == "NonGracefulTermination"
+		},
+		Handler: func(f *framework.Framework, ev *corev1.Event) {
+			framework.Failf("kube-apiserver reports a non-graceful termination: %#v", ev)
+		},
+	}
+	m.Run(ctx, f)
+}
+
+// LateConnectionsTest watches events late connections to the kube-apiserver during termination. These usually
+// mean that the load balancer do not configure their health probes correctly, by calling /readyz of the API
+// server (not /healthz !)
+type LateConnectionsTest struct{}
+
+func (LateConnectionsTest) Name() string { return "kube-apiserver-no-late-connections" }
+func (LateConnectionsTest) DisplayName() string {
+	return "LBs stop connecting to kube-apiserver during termination in time"
+}
+func (LateConnectionsTest) Setup(f *framework.Framework)    {}
+func (LateConnectionsTest) Teardown(f *framework.Framework) {}
+func (t *LateConnectionsTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		defer cancel()
+		<-done
+	}()
+
+	m := EventMonitor{
+		Namespace: "openshift-kube-apiserver",
+		Filter: func(ev *corev1.Event) bool {
+			return ev.Reason == "LateConnections"
+		},
+		Handler: func(f *framework.Framework, ev *corev1.Event) {
+			framework.Failf("kube-apiserver got connections late during termination, probably due to broken LB configuration: %#v", ev)
```

Comment:
It could be worth adding what consequences it could have on running tests (connection refused?) and what to do with this failure (Assign to API server? Networking? Blame the cloud and ignore?)

Reply:
:+1:

added